tsl/src/compression/api.c

```diff
@@ -334,19 +351,59 @@ decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_
 	/* acquire locks on src and compress hypertable and src chunk */
 	LockRelationOid(uncompressed_hypertable->main_table_relid, AccessShareLock);
 	LockRelationOid(compressed_hypertable->main_table_relid, AccessShareLock);
-	LockRelationOid(uncompressed_chunk->table_id, AccessShareLock); /*upgrade when needed */
 
-	/* aquire locks on catalog tables to keep till end of txn */
+	/*
+	 * Acquire an ExclusiveLock on the uncompressed and the compressed
+	 * chunk (the chunks can still be accessed by reads).
+	 *
+	 * The lock on the compressed chunk is needed because it gets deleted
+	 * after decompression. The lock on the uncompressed chunk is needed
+	 * to avoid deadlocks (e.g., caused by later lock upgrades or parallel
+	 * started chunk compressions).
+	 *
+	 * Note: Also the function decompress_chunk() will request an
+	 *       ExclusiveLock on the compressed and on the uncompressed
+	 *       chunk. See the comments in function about the concurrency of
+	 *       operations.
+	 */
+	LockRelationOid(uncompressed_chunk->table_id, ExclusiveLock);
+	LockRelationOid(compressed_chunk->table_id, ExclusiveLock);
+
+	/* acquire locks on catalog tables to keep till end of txn */
 	LockRelationOid(catalog_get_table_id(ts_catalog_get(), HYPERTABLE_COMPRESSION),
 					AccessShareLock);
+
 	LockRelationOid(catalog_get_table_id(ts_catalog_get(), CHUNK), RowExclusiveLock);
 
+	DEBUG_WAITPOINT("decompress_chunk_impl_start");
+
+	/*
+	 * Re-read the state of the chunk after all locks have been acquired and ensure
+	 * it is still compressed. Another process running in parallel might have
+	 * already performed the decompression while we were waiting for the locks to be
```

Comment:
Hmm i dont think thats actually possible since a decompress removes the compressed chunk so the wait for locks on it should fail.

Reply:
I was facing this situation. I have not found any check that prevents calling `LockRelationOid` with an `Oid` of a non-existent table. So, from my understanding, holding a lock of an Oid does not ensure that the table with this Oid actually exists.  If we do not perform this check, the operation later fails with something like:

```
ERROR:  could not open relation with OID 17015
```

But I prefer the clearer error message that is generated by this check.